src/datasets/utils/patching.py

```diff
@@ -39,19 +42,50 @@ def __init__(self, obj, target: str, new, attrs=None):
         self.target = target
         self.new = new
         self.key = target.split(".")[0]
-        self.original = getattr(obj, self.key, None)
+        self.original = {}
         self.attrs = attrs or []
 
     def __enter__(self):
-        *submodules, attr = self.target.split(".")
-        current = self.obj
-        for key in submodules:
-            setattr(current, key, _PatchedModuleObj(getattr(current, key, None), attrs=self.attrs))
-            current = getattr(current, key)
-        setattr(current, attr, self.new)
+        *submodules, target_attr = self.target.split(".")
+        # Patch modules:
+        # it's used to patch attributes of submodules like "os.path.join";
+        # in this case we need to patch "os" and "os.path"
+        for i in range(len(submodules)):
+            submodule = import_module(".".join(submodules[: i + 1]))
+            for attr in self.obj.__dir__():
+                obj_attr = getattr(self.obj, attr)
+                if obj_attr is submodule or (
+                    (isinstance(obj_attr, _PatchedModuleObj) and obj_attr._original_module is submodule)
+                ):
```

Comment:
A comment here would also be nice to explain that this allows us to patch renamed modules.